pkg/s3store/s3store.go

```diff
@@ -388,12 +392,21 @@ func (upload s3Upload) WriteChunk(ctx context.Context, offset int64, src io.Read
 	doneChan := make(chan struct{})
 	defer close(doneChan)
 
-	chunkProducer := s3ChunkProducer{
+	// If we panic or return while there are still files in the channel, then
+	// we may leak file descriptors. Let's ensure that those are cleaned up.
+	defer func() {
+		for file := range fileChan {
+			os.Remove(file.Name())
+			file.Close()
```

Comment:
Closing the file descriptor after removing the file should be fine on Linux. But, IIRC, on Windows you need to close all file descriptors before you can remove the file. So please change the order of these two lines, unless I am missing something.

FYI, deferred functions call are execute in the opposite order they are declared: https://tour.golang.org/flowcontrol/13. That's why we call `defer os.Remove` before `defer file.Close`

Reply:
Right, good catch. In each place where we do this file cleanup, I was thinking about using this same closure pattern instead of using two separate defers. Something like:

```go
defer func() {
  file.Close()
  os.Remove(file)
}()
```

Then we don't need to remember defer's stack behavior when we read this code. WDYT?

Reply:
Good idea. Maybe we can also refactor the file.Close and os.Remove into its own function.